File: wrappers/uniffi/swift/Tests/EqusSdkTests/KeyHandle.swift
import Testing
import Foundation
import Swifter
@testable import EqusSdk

@Suite(.serialized) class KeyHandleTests {


    @Test func getAlg() async throws {
        
        let keyHandle = self.mockKeyHandler()
        let alg = keyHandle.alg()

        #expect(EqusSdk.Alg.es256 == alg)
    }

    @Test func getPublicKey() async throws {
        
        let keyHandle = self.mockKeyHandler()
        let pubKey = try keyHandle.pubKey()

        #expect("testKey".data(using: .utf8)! == pubKey)
    } 

    @Test func getJwk() async throws {
        
        let keyHandle = self.mockKeyHandler()
        let jwk = keyHandle.jwk()

        #expect("{\"crv\":\"P-256\",\"kid\":\"618d228e-4767-4aa2-8683-c35c86d7025c\",\"kty\":\"EC\",\"x\":\"huX4QOwcvioB2N3njNOnTOtElUvf7KIQnm6NvdfK2bs\",\"y\":\"4qWecmcxVAXxyCBYuzxSpVRG7ETk9mO3RjUzsFUtDCg\"}" == jwk)
    }

    @Test func sign() async throws {
        let payload = "testPayload".data(using: .utf8)!
        
        let keyHandle = self.mockKeyHandler()
        let result = try await keyHandle.sign(payload: payload)

        #expect(result == payload)
    }

	@Test func verify() async throws {
        let keyHandle = self.mockKeyHandler()
        try await keyHandle.verify(data: "testData".data(using: .utf8)!, signature: "testSignature".data(using: .utf8)!)
	}
    
    private func mockKeyHandler() -> EqusSdk.KeyHandle {
        return wrapKeyHandleForTests(keyHandle: MockKeyHandle())
    }

}


final class MockKeyHandle : EqusSdk.KeyHandle {
    func pubKey() throws -> Data {
        return "testKey".data(using: .utf8)!
    }

    func jwk() -> String? {
        return "{\"crv\":\"P-256\",\"kid\":\"618d228e-4767-4aa2-8683-c35c86d7025c\",\"kty\":\"EC\",\"x\":\"huX4QOwcvioB2N3njNOnTOtElUvf7KIQnm6NvdfK2bs\",\"y\":\"4qWecmcxVAXxyCBYuzxSpVRG7ETk9mO3RjUzsFUtDCg\"}"
    }

    func alg() -> EqusSdk.Alg {
        return EqusSdk.Alg.es256
    }

    func sign(payload: Data) async throws -> Data {
        return payload
    }

    func verify(data: Data, signature: Data) async throws {
        #expect(data == "testData".data(using: .utf8)!)
        #expect(signature == "testSignature".data(using: .utf8)!)
    }
}

File: wrappers/uniffi/kotlin/src/test/kotlin/com/equs/KeyHandle.kt
package com.equs

import com.equs.credentials.Alg
import com.equs.credentials.KeyHandle as EqusSdkKeyHandle
import com.equs.credentials.setJniLibPath
import com.equs.credentials.wrapKeyHandleForTests
import kotlinx.coroutines.test.runTest
import org.junit.jupiter.api.BeforeAll
import kotlin.test.Test
import kotlin.test.assertEquals

class KeyHandle {
    companion object {
        @JvmStatic
        @BeforeAll
        fun setup() {
            setJniLibPath()
        }
    }

    @Test
    fun getAlg() = runTest {
        val keyHandle = mockKeyHandle()
        val alg = keyHandle.alg()

        assertEquals(Alg.ES256, alg)
    }

    @Test
    fun getPublicKey() = runTest {
        val keyHandle = mockKeyHandle()
        val pubKey = keyHandle.pubKey()

        assertEquals("testKey", String(pubKey, Charsets.UTF_8))
    }

    @Test
    fun getJwk() = runTest {
        val keyHandle = mockKeyHandle()
        val jwk = keyHandle.jwk()

        assertEquals(
            "{\"crv\":\"P-256\",\"kid\":\"618d228e-4767-4aa2-8683-c35c86d7025c\",\"kty\":\"EC\",\"x\":\"huX4QOwcvioB2N3njNOnTOtElUvf7KIQnm6NvdfK2bs\",\"y\":\"4qWecmcxVAXxyCBYuzxSpVRG7ETk9mO3RjUzsFUtDCg\"}",
            jwk,
        )
    }

    @Test
    fun sign() = runTest {
        val keyHandle = mockKeyHandle()
        val payload = "testPayload".toByteArray()
        val result = keyHandle.sign(payload)

        assertEquals(String(payload, Charsets.UTF_8), String(result, Charsets.UTF_8))
    }

    @Test
    fun verify() = runTest {
        val keyHandle = mockKeyHandle()
        val data = "testData".toByteArray()
        val signature = "testSignature".toByteArray()
        keyHandle.verify(data, signature)
    }

    private fun mockKeyHandle(): EqusSdkKeyHandle {
        return wrapKeyHandleForTests(MockKeyHandle())
    }
}

class MockKeyHandle : EqusSdkKeyHandle {
    override fun pubKey(): ByteArray {
        return "testKey".toByteArray()
    }

    override fun jwk(): String? {
        return "{\"crv\":\"P-256\",\"kid\":\"618d228e-4767-4aa2-8683-c35c86d7025c\",\"kty\":\"EC\",\"x\":\"huX4QOwcvioB2N3njNOnTOtElUvf7KIQnm6NvdfK2bs\",\"y\":\"4qWecmcxVAXxyCBYuzxSpVRG7ETk9mO3RjUzsFUtDCg\"}"
    }

    override fun alg(): Alg {
        return Alg.ES256
    }

    override suspend fun sign(payload: ByteArray): ByteArray {
        return payload
    }

    override suspend fun verify(data: ByteArray, signature: ByteArray) {
        assertEquals(String(data, Charsets.UTF_8), "testData")
        assertEquals(String(signature, Charsets.UTF_8), "testSignature")
    }
}